=str #19591 updated 'partition to three subscribers' to use Sink.seq …

israel opened the pull request:
…and added a new test to cover a case where partitioner return an index out of bounds

newca12 committed:
=htp fix multipart/form-data unmarshaller when no Content-Type field is present

RFC 2046:
… multipart/* …
If no Content-Type field is present it is assumed to be "message/rfc822" in a "multipart/digest" and "text/plain" otherwise

RFC 2388:
As with all multipart MIME types, each part has an optional "Content-Type", which defaults to text/plain.

Adebski committed:
=clu #19017 Race-condition issue when handling Topic/Group actors.

* Added reusable PerGroupingBuffer trait for pubsub implementation
* Moved mkKey methods to the Internal object
* Introduced passivate-like protocol between DistributedPubSubMediator/Topic and Topic/Group actors, contained in ChildActorTerminationProtocol messages.

briantopping committed:
=htp #19331 Make FileAndResourceDirectives work in OSGi environments

briantopping committed:
=osgi #19462 Fix version ranges

agolubev committed:
+str #18565 termination notification combinator

johanandren committed:
=doc #19315 docs for websocket client

ktoso committed:
!htc #19514,#18958 make config in HTTP non-case class + from reference

rkuhn committed:
#19440 replace Scala Future usage with CompletionStage in javadsl

This entails:

  * adding akka.pattern.PatternCS.* to enable ask etc. with
    CompletionStage
  * changing RequestContext to offer an ExecutionContextExecutor for the
    CompletionStage.*Async combinators
  * splitting up akka.stream.Queue for JavaDSL consistency

johanandren committed:
=str #19552 Replacing Conflate with Batch

ktoso committed:
!htc #19514 more case class => class boilerplate

ataraxer committed:
+str #19528 Add `propagateSubstreamCancel` flag to `splitWhen` and `splitAfter`

ataraxer committed:
=htc #19528 OutgoingConnectionBlueprint: replace splitWhen/prefixAndTail with custom stage

johanandren committed:
Cover sources and sinks as well, and change into a format that allows for toc

Commit:
19455 Simplify conflate signature for the common case

2m committed:
wip #19441 convert from Option to Optional in javadsl

ktoso committed:
!htc #19514 add withXYZ methods to all configs

johanandren committed:
=htt Test coverage for the Java HTTP extension

m-sp committed:
+str #19589 Add autoFlush option to  OutputStream Sinks

israel committed:
=str #19591 updated 'partition to three subscribers' to use Sink.seq and added a new test to cover a case where partitioner return an index out of bounds

israel committed:
=str #19591 updated 'partition to three subscribers' to use Sink.seq and added a new test to cover a case where partitioner return an index out of bounds

ochrons committed:
Fixes #19593 a potential race condition in FileSubscriber.
If an error occurs in the stream, the completionPromise is completed
before the file is closed, leading to potential race condition if the
file is then tried to be deleted upon the promise completion, while it's
still open.

Commit:
=str adding documentation for batch & batchAggregated combinators

raboof committed:
=doc fix comment for safe-shutdown-example-bad

The `PoisonPill` of course isn't stashed - probably a copy-paste-leftover from the 'good' example.

hepin1989 commented on akka-stream/src/main/scala/akka/stream/scaladsl/Graph.scala:
```diff
@@ -469,6 +471,8 @@ final class Broadcast[T](private val outputPorts: Int, eagerCancel: Boolean) ext
 
 object Partition {
 
+  case class PartitionOutOfBoundsException(msg:String) extends RuntimeException(msg) with NoStackTrace
```
it's not a good idea for case class.

israel replied:
what is not a good idea? the "case" part?
I "took" it from other places in the project. for example:
https://github.com/akka/akka/blob/a12451e007d2d40564860f5a6a1a675ded8824d5/akka-stream/src/main/scala/akka/stream/scaladsl/One2OneBidiFlow.scala#L14

hepin1989 replied:
A normal class is OK I think.take a look at other part?

israel replied:
what part? I am sorry, I don't understand

drewhk replied:
what is wrong with a normal IndexOutOfBoundsException?

israel replied:
It does not compare well in https://github.com/israel/akka/blob/israel-%2319591/akka-stream-tests/src/test/scala/akka/stream/scaladsl/GraphPartitionSpec.scala#L179

israel replied:
```
scala> val x = new java.lang.IndexOutOfBoundsException("123")
x: IndexOutOfBoundsException = java.lang.IndexOutOfBoundsException: 123

scala> val y = new java.lang.IndexOutOfBoundsException("123")
y: IndexOutOfBoundsException = java.lang.IndexOutOfBoundsException: 123

scala> x == y
res0: Boolean = false

scala> case class MyException(msg:String) extends java.lang.RuntimeException(msg)
defined class MyException

scala> val a = MyException("123")
a: MyException = MyException: 123

scala> val b = MyException("123")
b: MyException = MyException: 123

scala> a == b
res1: Boolean = true

scala> 
```

drewhk replied:
Just use `expectError()` which return the Throwable, then you can look at the message or call an `isInstanceOf`

rkuhn replied:
:+1:

rkuhn committed:
Merge pull request #19578 from akka/wip-19440-completionStage-RK

 #19440 replace Scala Future usage with CompletionStage in javadsl

prayagverma committed:
=pro #19068 All copyrights ranges and single years updated to a range ending in 2016

Commit:
Merge pull request #19498 from agolubev/agolubev-#18565-TerminationNotification-combinator

+str #18565 termination notification combinator

rkuhn committed:
convert watchTermination to CompletionStage for Java

viktorklang committed:
=str - Sprinkles some finals in Tcp and improves docs in Flow

johanandren committed:
Merge pull request #19603 from akka/wip-19068-2016-copyrights-johanandren

=pro #19068 All copyrights ranges updated with 2016

johanandren committed:
!htt #19514 Update RouterSettings to be non case class

rkuhn committed:
remove experimental marker from stream/testkit/parsing/http-core

johanandren committed:
Merge pull request #19604 from akka/wip-19441-convert-from-Option-to-Optional-in-javadsl-johanandren

!str #19441 convert from Option to Optional in javadsl

johanandren committed:
Merge pull request #19596 from hochgi/hochgi-adding-batch-to-docs

=str adding documentation for batch & batchWeighted combinators

johanandren committed:
Merge pull request #19590 from m-sp/master

+str #19589 Add autoFlush option to  OutputStream Sinks

johanandren committed:
Merge pull request #19607 from akka/wip-stream-non-experimental-RK

remove experimental marker from stream/testkit/parsing/http-core

johanandren committed:
Merge pull request #19606 from akka/wip-19514-making.config-classes-easier-to-evolve-bin-compat-johanandren

!htc #19514 make config classes easier to evolve bin-compat (rebase)

danxmoran committed:
=pro #19581 Ignore MiMa changes in validatePullRequest.

johanandren committed:
+htt #19618 adding missing factory methods for server settings in Java API

agolubev committed:
=str #19585 fix long initialization for InputStreamSinkSpec

johanandren committed:
!str #19620 Java API for shutting connection pool down

johanandren committed:
=doc #16262 temporary disable failing udp multicast doc test

johanandren committed:
Merge pull request #19630 from akka/wip-16262-mark-as-pending-johanandren

=doc #16262 temporary disable failing udp multicast doc test

johanandren committed:
=act #19616 removing unused setting received-message-size-limit for UDP

patriknw committed:
Merge pull request #19154 from Adebski/Adebski-distributed-pub-sub-race-condition-fix

Race-condition problem in distributed-pub-sub fixed by introducing passivate-like protocol for mediator and Topic actors.

naferx committed:
Update copyright year

johanandren committed:
=htt #19623 Early response no longer causing push or pull twice errors

notxcain committed:
!htt #19628 Make CORS response headers extend ResponseHeader

johanandren committed:
Merge pull request #19629 from notxcain/fix-19628

Make necessary CORS headers extend ResponseHeader

johanandren committed:
=htt #19623 optimizations and more careful handling of complete and failure in PrepareRequest

rkuhn committed:
#19637 fix BC issue in AsyncWriteJournal

johanandren committed:
=str #19469 More verbose logging of stage and tcp errors

* Logs all exceptions thrown in GraphStages at error level
* Logs any exception leading to an abort of a tcp connection at debug level

johanandren committed:
Merge pull request #19598 from raboof/patch-3

=doc fix comment for safe-shutdown-example-bad

rkuhn committed:
Merge pull request #19644 from akka/wip-19637-fix-BC-AsyncWriteJournal-RK

briantopping committed:
Merge remote-tracking branch 'upstream/master' into resources-with-osgi

rkuhn committed:
fix ActorRefWithAck when buffer has run full #19653

rkuhn committed:
Merge pull request #19638 from naferx/copyright-year

Update copyright year

jypma committed:
+htt #19657 Fix type mapping to re-enable Java HTTPS servers

Commit:
Merge pull request #19654 from akka/wip-19653-ActorRefWithAck-RK

fix ActorRefWithAck when buffer has run full #19653

Commit:
Merge pull request #19640 from akka/wip-19623-early-response-errors-johanandren

=htt #19623 Early response no longer causing push or pull twice errors

viktorklang committed:
=str - Minor improvements to Framing

Commit:
Merge pull request #19656 from jypma/java_https_fix

[Java DSL] Fix type mapping to re-enable Java HTTPS servers

Commit:
#19392: asInputStream timeout

drewhk commented:
This is not super urgent, but I think this should use `IndexOutOfBoundsException`. I closing it for now since the PR queue is filling up, but please reopen once there is something new.

Commit:
Merge pull request #19605 from akka/wip-minor-touchups-3000-√

=str - Sprinkles some finals in Tcp and improves docs in Flow

Commit:
Merge pull request #19652 from briantopping/wip-19462

OSGI bundle changes.

johanandren committed:
Merge pull request #19586 from akka/wip-cover-java-http-interface-with-tests-johanandren

Java HTTP Extension test coverage

Commit:
Merge pull request #19594 from ochrons/fix-race-condition-filesubscriber

+str #19593 Fix a potential race condition in FileSubscriber.

Commit:
Merge pull request #19631 from akka/wip-19616-remove-unused-udp-setting-johanandren

=act #19616 removing unused setting received-message-size-limit for UDP

drewhk commented:
OK TO TEST

Commit:
Merge pull request #19660 from drewhk/wip-19392-fix-asinputstream-drewhk

#19392: asInputStream timeout

Commit:
Merge pull request #19647 from akka/wip-19469-better-stream-error-logging-johanandren

=str #19469 Log all graph stage exceptions

johanandren committed:
=str #19664 Docs no longer claim that framer delimiter is included in frames

Commit:
Merge pull request #19666 from akka/wip-19664-incorrect-framing-docs-johanandren

=str #19664 Docs no longer claim that framer delimiter is included

rkuhn committed:
remove verbose expected stacktrace from JournalSpec

talpr committed:
=clu #19622 Use full address in ClusterShardingStats

Commit:
Don't leak FunctionRef instances

Commit:
Merge pull request #19659 from akka/wip-minor-framing-touchups-√

=str - Minor improvements to Framing

Commit:
Merge pull request #19669 from drewhk/wip-functionref-memleak-drewhk

Don't leak FunctionRef instances

2m committed:
Add support for OpenJDK to release script

2m committed:
Merge pull request #19671 from 2m/wip-release-script-openjdk

Add support for OpenJDK to release script

retronym committed:
!clu #19679 make MemberStatus a sealed class

My assumption is that it the absence of the sealed modifier
was an oversight. Marking it as sealed will avoid exhausitity
warnings from upcoming Scala compiler version in `highestPriorityOf`.

johanandren committed:
!htt #19674 Audit HTTP headers and mark correctly as response or request

johanandren committed:
=hts #19663 Bind-timeout cannot be infinite, reference.conf corrected

johanandren committed:
Merge pull request #19667 from akka/wip-nicer-JournalSpec-RK

remove verbose expected stacktrace from JournalSpec

Commit:
#19553 One2OneBidi should report truncation also if wrapped flow cancels early

johanandren committed:
Merge pull request #19686 from johanandren/wip-19663-bind-timeout-cannot-be-infinite-johanandren

=hts #19663 Bind-timeout cannot be infinite, reference.conf corrected

johanandren committed:
Merge pull request #19668 from talp-traiana/talp-19622-clustershardingstats-node-name-fix

=clu #19622 Use full address in ClusterShardingStats

2m committed:
Fix double border from pre sections in docs.

2m committed:
Merge pull request #19691 from 2m/wip-docs-style-fixes

Remove double border from pre sections in docs (forward port of #18352)

2m committed:
Merge pull request #19680 from retronym/topic/seal-member-status

!clu #19679 make MemberStatus a sealed class

Commit:
Merge pull request #19582 from drewhk/wip-19445-simplify-conflate-drewhk

#19455 Simplify conflate signature for the common case

naferx committed:
Remove unused imports

ataraxer committed:
!htp #19528 HttpServerBluePrint: close connection on request entity cancellation

keredin committed:
Fix memory leak in AddressTerminatedTopic

refs #19672

2m committed:
#19639 Supress warnings on custom headers

johanandren committed:
One small step for mankind but a pretty big one for the stage overview

johanandren committed:
Java and Scala overview separated. delay and initialDelay added

johanandren committed:
!str #19698 Remove tuple from Java API Source signatures

Commit:
=str - correct typo in 'Source.from()' function documentation

rkuhn committed:
prohibit concurrent write request from same persistenceId #19694

- also use scalactic.ConversionCheckedTripleEquals
- also remove usage of the setting `max-message-batch-size`

rkuhn committed:
add hook before each persistence TCK test case

rkuhn committed:
add maxFixedBufferSize materializer setting #19702

johanandren committed:
=htc #19692 Improved docs for the web socket message classes

johanandren committed:
Merge pull request #19685 from johanandren/wip-19674-http-header-request-or-response-audit-johanandren

!htt #19674 Audit HTTP headers and mark correctly as response or request

johanandren committed:
Merge pull request #19543 from akka/wip-19315-document-websocket-client-johanandren

=doc #19315 websocket client docs

rkuhn committed:
Merge pull request #19697 from 2m/wip-rendering-warning-2m

#19639 Supress warnings on custom headers

rkuhn committed:
Merge pull request #19699 from johanandren/wip-19698-scala-tuples-in-java-source-johanandren

!str #19698 Remove tuple from Java API Source signatures

johanandren committed:
Merge pull request #19706 from johanandren/wip-19692-ws-message-class-docs-johanandren

=htc #19692 Improved docs for the web socket message classes

ktoso committed:
Merge pull request #19695 from naferx/remove-unused-imports

=htt Remove unused imports

johanandren committed:
!str #19710 Use Java types in Attributes Java API

Commit:
Merge pull request #19625 from agolubev/agolubev-#19585-Failed-InputStreamSinkSpec

=str #19585 fix long initialization for InputStreamSinkSpec

Commit:
Merge pull request #19665 from dorin-cbz/master

Correct typo in 'Source.from()' function documentation

Commit:
Merge pull request #19564 from akka/wip-19552-implement-conflate-in-terms-of-batch-johanandren

=str #19552 Replacing Conflate with Batch

ktoso committed:
Merge pull request #19415 from briantopping/resources-with-osgi

=htp #19331 Make FileAndResourceDirectives work in OSGi environments

drewhk commented:
Failure was unrelated.

drewhk commented:
PLS BUILD

Commit:
Merge pull request #19641 from akka/wip-19547-make-stages-overview-complete-johanandren

=doc #19547 Better overview of stages

ktoso committed:
Merge pull request #19624 from danxmoran/danxmoran-19581-ignore-mima

=pro #19581 Ignore MiMa changes in validatePullRequest.

Commit:
Merge pull request #19696 from keredin/master

Fix memory leak in AddressTerminatedTopic, refs #19672

johanandren committed:
=htc #18993 HttpConnectionPool constructor made private as it leaks impl type

ktoso committed:
+htc #19708 provides default marshaller for akka.Done (200, empty text response)

johanandren committed:
Merge pull request #19712 from johanandren/wip-19699-no-private-types-in-http-public-api-johanandren

=htc #18993 HttpConnectionPool constructor made private as it leaks i…

patriknw committed:
silence deadletter logging of SubscribeAck

* since it is normally (practially always) a local
  messages send that will not be rejected you don't
  care about the ack

naferx committed:
Add Typesafe copyright header

johanandren committed:
=str #19721 Failing FlowConflateSpec and NPE in Batch fixed

ktoso committed:
Merge pull request #19718 from naferx/wip/copyright-headers

Add Typesafe copyright header

ktoso committed:
=doc hide auto downing from default sample config in cluster

Commit:
spelling corrections

Commit:
fix spelling errors in javadsl/Flow.scala

ktoso committed:
-doc remove outdated list of http frameworks built on Akka

ktoso committed:
Merge pull request #19727 from enelson/master

spelling corrections

ktoso committed:
Merge pull request #19722 from johanandren/wip-19721-flowconflatespec-failing-johanandren

=str #19721 Failing FlowConflateSpec and NPE in Batch fixed

fgeller committed:
Fixes typo in docstring.

2m committed:
#19719 add ThrottleMode Java API

johanandren committed:
Merge pull request #19717 from akka/wip-silence-sub-ack-patriknw

silence deadletter logging of SubscribeAck

johanandren committed:
Merge pull request #19724 from ktoso/wip-dontrecommendautodowning-ktoso

=doc hide auto downing from default sample config in cluster

ktoso committed:
Merge pull request #19715 from ktoso/wip-done-marshalling

+htc #19708 provides default marshaller for akka.Done (200, empty text…

johanandren committed:
Merge pull request #19673 from newca12/newca12-#19610-fix-multipart-form-data-unmarshaller

=htp fix multipart/form-data unmarshaller when no Content-Type field

johanandren commented:
PLS BUILD

ktoso committed:
Merge pull request #19528 from ataraxer/wip-http-response-cancellation

Http: HttpEntity stream cancellation

Commit:
Merge pull request #19730 from 2m/wip-#19719-throttle-mode-java-2m

#19719 add ThrottleMode Java API

aloiscochard committed:
19452: Contramap and fromFunction

viktorklang committed:
=str - removing superfluous try-catch in stream-customize.rst

Commit:
Merge pull request #19690 from drewhk/wip-19553-one2onebidi-cancellation-drewhk

#19533 One2OneBidi should report truncation also if wrapped flow cancels early

ktoso committed:
=doc #19737 explain websocket -> WebSocket change in migration guide

rkuhn committed:
Merge pull request #19700 from akka/wip-19694-journal-protocol-RK

prohibit concurrent write request from same persistenceId #19694

johanandren committed:
Merge pull request #19711 from johanandren/wip-19710-scala-types-attributes-java-api-johanandren

!str #19710 Use Java types in Attributes Java API

johanandren committed:
=htc #19741 cannot pull twice bug in OutgoingConnectionBlueprint fixed

rkuhn committed:
do not require partial batch replay in TCK #19728

2m committed:
Fix a couple of build issues.

* fix sample project names
* remove experimental tag from stream tests
* do not publish stream tck tests

rkuhn committed:
Merge pull request #19703 from akka/wip-19702-maxFixedBufferSize-RK

add maxFixedBufferSize materializer setting #19702

2m committed:
Merge pull request #19747 from 2m/wip-fix-multiple-build-issues-2m

Fix a couple of build issues

2m committed:
#19662 Reuse connection pool settings if none specified

Commit:
Merge pull request #19729 from fgeller/patch-1

Fixes typo in docstring.

Commit:
Merge pull request #19739 from ktoso/wip-cleanup-docs-add-webSocket-note-ktoso

Explain WebSocket name change and remove obsolete docs part

rkuhn committed:
Merge pull request #19744 from akka/wip-19728-TCK-atomic-RK

do not require partial batch replay in TCK #19728

Commit:
Merge pull request #19742 from johanandren/wip-fix-failing-HighLevelOutgoingConnectionSpec-johanandren

=htc #19741 cannot pull twice bug in OutgoingConnectionBlueprint fixed

rkuhn committed:
Merge pull request #19684 from 2m/wip-#19662-reuse-pool-settings-2m

#19662 Reuse connection pool settings if none specified

rkuhn committed:
Merge pull request #19736 from akka/wip-there-is-no-try-√

=str - removing superfluous try-catch in stream-customize.rst

rkuhn commented on akka-stream/src/main/scala/akka/stream/scaladsl/Graph.scala:
```diff
@@ -508,7 +512,7 @@ final class Partition[T](outputPorts: Int, partitioner: T ⇒ Int) extends Graph
         val idx = partitioner(elem)
         if (idx < 0 || idx >= outputPorts)
-          failStage(new IndexOutOfBoundsException(s"partitioner must return an index in the range [0,${outputPorts - 1}]. returned: [$idx] for input [$elem]."))
+          failStage(PartitionOutOfBoundsException(s"partitioner must return an index in the range [0,${outputPorts - 1}]. returned: [$idx] for input [$elem]."))
```
hmm, this should probably not include the element—we might want to perform an audit that ensure that we do not log `elem.tostring` anywhere (because that could actually be HUGE)

israel replied:
do you want to remove the elem print at all or maybe just print first N characters of it?

rkuhn replied:
just calling `toString` may be of arbitrary cost, so we typically just print the class name (which admittedly is of limited use here)

Commit:
Merge pull request #19734 from drewhk/wip-19452-contramap-drewhk

#19452: Contramap and fromFunction

israel committed:
=str #19591 updated 'partition to three subscribers' to use Sink.seq and added a new test to cover a case where partitioner return an index out of bounds

drewhk commented:
argh, something was wrong with your rebase. Too many commits here.

israel commented:
you're right. I will just create a new PR...

On Thu, Feb 11, 2016 at 9:44 PM drewhk <notifications@github.com> wrote:

> argh, something was wrong with your rebase. Too many commits here.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/akka/akka/pull/19592#issuecomment-183032050>.
>

israel commented:
I have opened a new corrected PR https://github.com/akka/akka/pull/19751

drewhk commented:
Ok, this one can be closed. Thanks @israel for your contribution!
